src/ir/local-graph.h

```diff
@@ -228,7 +232,51 @@ struct LazyLocalGraph : public LocalGraphBase {
     return *locations;
   }
 
+  // Query which gets a set influences (reaches), assuming that a given
+  // expression blocks the flow. The obstacle must be of the class
+  // obstacleClass that is provided in the constructor (so that we set up our
+  // data structures in a way that is ready for any item of that class). For
+  // example:
+  //
+  //  1. set
+  //  2. get
+  //  3. call
+  //  4. get
+  //
+  // If we use Call as obstacleClass, then we can use any call as an obstacle in
+  // a query, such as the call on line #3 here. Then:
+  //
+  //  getSetInfluencesGivenObstacle(#1, #3) => { #2 }
+  //
+  // That is, if the call (#3) obstructs the set (#1) then it only reaches the
+  // get on line #2, and is prevented from reaching line #4, since line #3 is in
+  // the way.
+  //
+  // Obstacle queries are a simple way to check what would happen if code were
+  // moved around. In the example above, imagine that we are considering moving
+  // the set from line #1 to #3, then we can simulate that by the query we just
+  // mentioned: originally we can reach both #2 and #4, and asking whether we
+  // can still do so if we moved to location #3 is equivalent to asking if #3
+  // blocks us from each of the original gets #2 & #4. That is, if #3 blocks us
+  // from all our gets then that means that #3 reaches them all (& if we were in
+  // the place of that obstacle, so would we). And that would be the case if #2
+  // did not exist in the first place, but given #2 in the full example, we can
+  // see that #2 is still reachable despite the obstacle, showing that moving
+  // the set would cause an issue.
+  //
+  // An alternative to such obstacle queries is to just modify BinaryenIR,
+  // construct a new LocalGraph, and see what gets it has (and whether they
+  // match the original ones). Obstacle queries are basically a more efficient
+  // way to do that, by reusing the same LocalGraph for all such queries. To do
+  // so we must decide which things can play the role of obstacle
+  // (obstacleClass), and we then place them in the LocalGraph's internal IR.
+  // A query for a particular obstacle is then easy and cheap.
+  SetInfluences getSetInfluencesGivenObstacle(LocalSet* set,
+                                              Expression* obstacle);
```

Comment:
I think it would be more straightforward to frame this in terms of reachability. Ultimately what we want to know is whether control flow reaches the same set of gets from some set and some arbitrary other instruction. Computing and comparing the set of reached gets is one possible implementation of that, but computing that second set is not ultimately the problem we want to solve, so it seems odd to have an API for it.

Can we have an API like `bool hasSameSetInfluences(LocalSet* set, Expression* expr)`? It would just require moving the comparison of the sets from the client to the utility.

Reply:
`hasSameSetInfluences(set, expr)` would need to either

1. Start from `expr` and flow forward to find the gets, which is usually less efficient (it will reach dead ranges at the end of the function, past any gets), or
2. Start from all gets of the relevant local index and flow backwards. This is less efficient than starting from a known set of gets, which is what the current PR does.

Overall the problem we have here is not fully general: we want to consider a small "tweak" to a `local.set`, not an arbitrary new position for it.

But I do see what you're saying that the API here is a little odd. Thinking now, maybe another option is `findGetsReachingExpr(gets, expr)` which would flow back from those gets specifically. It would be called using `findGetsReachingExpr(getSetInfluences(set), expr)` (that is, the gets are the original gets that the set reaches), but might have more general uses later too. This should be efficient as it flows back from the relevant gets, and it will stop at either `expr` or the `set` (we don't need to tell it the `set` explicitly, since it will stop at any set of that index). Thoughts?

Reply:
Hmm, maybe I'm misunderstanding something. My hope was that we could just inline some of the current client code into the utility to provide a simpler interface where the client just does one call. I wasn't thinking that the actual algorithm would have to change.

Reply:
Anything else I can clarify here?

Reply:
In discussion offline we agreed that an API like `findGetsReachedBy(gets, expr)` would make sense. However, mid-way into writing the code to switch to it I realized that doesn't quite work. Consider this:

```wat
(block $label
 ;; A struct.set that may be skipped by the br in the if.
 (struct.set $struct 0  ;; structSet
  (local.tee $struct  ;; localTee
   (struct.new_default $struct)
  )
  (if (result i32)
   (..)
   (then
    (br $label)
   )
   (else
    (i32.const 0)
   )
  )
 )
)
(drop
 (struct.get $struct 0
  (local.get $struct)  ;; get
 )
)
```
If we ask `findGetsReachedBy({ get }, structSet)` then we get `{ get }`  - the structSet reaches that get. But the br may skip over it! So while the structSet and the localTee can both reach the get, one of them might not be passed by on the way. That is, we can go straight from the localTee to the get, skipping the structSet.

And that is the difference we care about here. If the localTee were moved to the structSet, and then skipped, then the get would read a different value from the local (whatever was in the local before all this code).

To compute this "must pass by X on the way from A to B" we do need both the gets, and the set, and the obstacle along the way, I think. So an "obstacle query" is still the best idea I can think of.

Reply:
Can we do `findGetsDominatedByExpr`?

Reply:
Good idea, yeah, that might work. "Dominated" here meaning "is the only set for them?" (so it dominates in the CFG sense, and also no other set is between them). I'll write that up.

Reply:
After more offline discussion we agreed to rename the API to "canMoveSet", which goes directly to the point of what we are computing, keeping the API itself simple and easy to explain. The implementation does do an obstacle analysis to compute things, but that is an internal detail.